yarn-project/sequencer-client/src/sequencer/abstract_phase_manager.ts

```diff
@@ -377,165 +383,107 @@ export abstract class AbstractPhaseManager {
-      revertiblePublicDataReads.find(item => item.leafSlot.equals(read.leafSlot) && item.value.equals(read.value)),
-    );
-    const simRevertiblePublicDataUpdateRequests = simPublicDataUpdateRequests.filter(update =>
-      revertiblePublicDataUpdateRequests.find(
-        item => item.leafSlot.equals(update.leafSlot) && item.newValue.equals(update.newValue),
-      ),
-    );
-
-    const simNonRevertiblePublicDataReads = simPublicDataReads.filter(read =>
-      nonRevertiblePublicDataReads.find(item => item.leafSlot.equals(read.leafSlot) && item.value.equals(read.value)),
-    );
-    const simNonRevertiblePublicDataUpdateRequests = simPublicDataUpdateRequests.filter(update =>
-      nonRevertiblePublicDataUpdateRequests.find(
-        item => item.leafSlot.equals(update.leafSlot) && item.newValue.equals(update.newValue),
-      ),
-    );
-
-    // Assume that kernel public inputs has the right number of items.
-    // We only want to reorder the items from the public inputs of the
-    // most recently processed top/enqueued call.
-    const numRevertibleReadsInKernel = arrayNonEmptyLength(publicInputs.end.publicDataReads, f => f.isEmpty());
-    const numRevertibleUpdatesInKernel = arrayNonEmptyLength(publicInputs.end.publicDataUpdateRequests, f =>
-      f.isEmpty(),
-    );
-    const numNonRevertibleReadsInKernel = arrayNonEmptyLength(publicInputs.endNonRevertibleData.publicDataReads, f =>
-      f.isEmpty(),
-    );
-    const numNonRevertibleUpdatesInKernel = arrayNonEmptyLength(
-      publicInputs.endNonRevertibleData.publicDataUpdateRequests,
-      f => f.isEmpty(),
-    );
-
-    // Validate all items in enqueued public calls are in the kernel emitted stack
-    const readsAreEqual =
-      simRevertiblePublicDataReads.length + simNonRevertiblePublicDataReads.length === simPublicDataReads.length;
-
-    const updatesAreEqual =
-      simRevertiblePublicDataUpdateRequests.length + simNonRevertiblePublicDataUpdateRequests.length ===
-      simPublicDataUpdateRequests.length;
+function removeRedundantPublicDataWrites(publicInputs: PublicKernelCircuitPublicInputs) {
+  const patch = <N extends number>(requests: Tuple<PublicDataUpdateRequest, N>) => {
+    const lastWritesMap = new Map<string, PublicDataUpdateRequest>();
+    for (const write of requests) {
+      const key = write.leafSlot.toString();
+      lastWritesMap.set(key, write);
+    }
+    return requests.filter(write => lastWritesMap.get(write.leafSlot.toString())?.equals(write));
+  };
+
+  publicInputs.end.publicDataUpdateRequests = padArrayEnd(
+    patch(publicInputs.end.publicDataUpdateRequests),
+    PublicDataUpdateRequest.empty(),
+    MAX_REVERTIBLE_PUBLIC_DATA_UPDATE_REQUESTS_PER_TX,
+  );
+
+  publicInputs.endNonRevertibleData.publicDataUpdateRequests = padArrayEnd(
+    patch(publicInputs.endNonRevertibleData.publicDataUpdateRequests),
+    PublicDataUpdateRequest.empty(),
+    MAX_NON_REVERTIBLE_PUBLIC_DATA_UPDATE_REQUESTS_PER_TX,
+  );
```

Comment:
I think we need to convince the kernel that we squashed thing correctly here?

Reply:
Yes, we will eventually absolutely need to do that. This is piggybacking on the assumption that we don't do that currently, but will need to eventually.